File: 2021/lib/day03.rb
# https://adventofcode.com/2021/day/3

class Day03

  MORE_COMMON = -> (count) { count["0"] > count["1"] ? "0" : "1" }
  LESS_COMMON = -> (count) { count["0"] > count["1"] ? "1" : "0" }

  class Report1
    def initialize(input)
      number_size = input[0].size
      @counts = []
      number_size.times do |i|
        @counts << input.map { |n| n[i] }.tally
      end
    end

    def count(fxn)
      @counts
        .map { |count| fxn.call(count) }
        .join("").to_i(2)
    end

    def epsilon
      count MORE_COMMON
    end

    def gamma
      count LESS_COMMON
    end
  end

  def self.part1(input)
    report = Report1.new input
    return [report.gamma, report.epsilon]
  end

  class Report2
    def initialize(input)
      @input = input
    end

    def count_filter(fxn)
      numbers = @input
      numbers.first.size.times do |i|
        count = numbers.map { |n| n[i] }.tally
        target = fxn.call(count)
        numbers = numbers.select { |n| n[i] == target }
        break if numbers.size == 1
      end
      numbers.first.to_i(2)
    end

    def oxygen
      count_filter MORE_COMMON
    end

    def co2
      count_filter LESS_COMMON
    end
  end

  def self.part2(input)
    report = Report2.new input
    return [report.oxygen, report.co2]
  end
end

File: 2021/lib/day03.exs
# Answer based on Jose Valim's stream
# I'm learning elixir

use Bitwise, only_operators: true

{_, input} = File.read("input/day03.txt")

numbers =
  input
  |> String.split("\n", trim: true)
  |> Enum.map(&(&1 |> String.to_charlist() |> List.to_tuple()))

[sample | _] = numbers
digit_count = tuple_size(sample)
half = div(length(numbers), 2)

gamma =
  for i <- 0..(digit_count - 1) do
    ones = Enum.count(numbers, &(elem(&1, i) == ?1))
    if ones > half, do: ?1, else: ?0
  end
  |> List.to_integer(2)

bitmask = (1 <<< digit_count) - 1
epsilon = ~~~gamma &&& bitmask
part1 = gamma * epsilon

defmodule Recursion do
  defp recur([number], _pos, _fx) do
    number
    |> Tuple.to_list()
    |> List.to_integer(2)
  end

  defp recur(numbers, pos, fx) do
    ones = Enum.count(numbers, &(elem(&1, pos) == ?1))
    zeroes = length(numbers) - ones
    target = fx.(zeroes, ones)
    numbers = Enum.filter(numbers, &(elem(&1, pos) == target))
    recur(numbers, pos + 1, fx)
  end

  def o2(numbers) do
    recur(numbers, 0, fn zeroes, ones ->
      if zeroes > ones, do: ?0, else: ?1
    end)
  end

  def co2(numbers) do
    recur(numbers, 0, fn zeroes, ones ->
      if zeroes > ones, do: ?1, else: ?0
    end)
  end
end

part2 = Recursion.o2(numbers) * Recursion.co2(numbers)

IO.write("Part 1: ")
IO.puts(part1)
IO.write("Part 2: ")
IO.puts(part2)
